### PR TITLE
Optimize `mk_region`

### PR DESCRIPTION
**Pull request by nnethercote**
PR #107869 avoiding some interning under `mk_ty` by special-casing `Ty` variants with simple (integer) bodies. This PR does something similar for regions.

r? @compiler-errors

### CODE REVIEW

File: compiler/rustc_mir_transform/src/large_enums.rs [hunk 1]
```diff
@@ -141,10 +141,7 @@ impl EnumSizeOpt {
                         self.candidate(tcx, param_env, ty, &mut alloc_cache)?;
                     let alloc = tcx.global_alloc(alloc_id).unwrap_memory();
 
-                    let tmp_ty = tcx.mk_ty(ty::Array(
-                        tcx.types.usize,
-                        Const::from_usize(tcx, num_variants as u64),
-                    ));
+                    let tmp_ty = tcx.mk_array(tcx.types.usize, num_variants as u64);
```
Comment (compiler-errors) [inline compiler/rustc_mir_transform/src/large_enums.rs:144]: lol I caught this one earlier today too :)

**Review by compiler-errors — APPROVED** [review]
r=me when perf comes back green, looks good to me.

edit: oh, also seems to need a rebase.

**Comment by nnethercote** [comment]
Thanks for the fast review!

@bors try @rust-timer queue

**Comment by nnethercote** [comment]
BTW, I tried similar things with `mk_predicate`, but predicates are more complex. There are no variants containing just an integer, for example. The existing `reuse_or_mk_predicate` function is effective, I couldn't improve on that.

**Comment by nnethercote** [comment]
I rebased.

@bors r=compiler-errors

**Pull request merged**